File: contest/LeetCode_Weekly_Contest_206/Min_Cost_to_Connect_All_Points.java
class Solution {
    class Edge {
        int u, v, w;
    };

    int n = 0;
    Edge[] e = new Edge[1000005];
    int[] p = new int[1005];

    int find(int u) {
        if (u == p[u]) return u;
        p[u] = find(p[u]);
        return p[u];
    }
    public int minCostConnectPoints(int[][] points) {
        n = points.length;

        int k = 0;
        for (int i = 0; i < n; ++i)
            for (int j = i+1; j < n; ++j) {
                int xi = points[i][0], yi = points[i][1];
                int xj = points[j][0], yj = points[j][1];
                int dist = Math.abs(xi - xj) + Math.abs(yi - yj);
                e[k] = new Edge();
                e[k].u = i;
                e[k].v = j;
                e[k].w = dist;
                k++;
            }

        Arrays.sort(e, 0, k, new Comparator<Edge>() {
            public int compare(Edge o1, Edge o2) {
                return o1.w - o2.w;
            }
        });

        for (int i = 0; i < n; ++i) p[i] = i;

        int ret = 0;
        for (int i = 0; i < k; ++i) {
            int u = e[i].u,  v = e[i].v, w = e[i].w;
            int pu = find(u);
            int pv = find(v);
            if (pu == pv) continue;
            p[pu] = pv;
            ret += w;
        }
        return ret;
    }
}

File: contest/LeetCode_Weekly_Contest_206/Min_Cost_to_Connect_All_Points.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function(points) {
    var n = points.length;
    var p = []
    for (var i = 0; i < n; i++) p[i] = i
    var e = []
    var m = 0
    for (var i = 0; i < n; i++)
        for (var j = i+1; j < n; j++) {
            var dist = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1])
            e[m] = [i, j, dist]
            m++
        }
    
    e.sort(function (a, b) {return a[2] - b[2]})
    
    var find = function (x) {
        if (x == p[x]) return x
        p[x] = find(p[x])
        return p[x]
    }
    
    var ret = 0
    for (var i = 0; i < m; i++) {
        var pu = find(e[i][0])
        var pv = find(e[i][1]);
        if (pu == pv) continue;
        p[pu] = pv
        ret += e[i][2]
    }
    return ret
};
